arwen/host/execution_async_context.go

```diff
@@ -0,0 +1,435 @@
+	syncExecutionOnly bool,
+) error {
+	for _, asyncCall := range asyncCallGroup.AsyncCalls {
+		err := host.executeAsyncCall(asyncCall, syncExecutionOnly)
+		if err != nil {
+			return err
+		}
+	}
+
+	asyncCallGroup.DeleteCompletedAsyncCalls()
+
+	// If ALL the AsyncCalls in the AsyncCallGroup were executed synchronously,
+	// then the AsyncCallGroup can have its callback executed.
+	if asyncCallGroup.IsCompleted() {
+		// TODO reenable this, after allowing a gas limit for it and deciding what
+		// arguments it receives (this method is currently a NOP and returns nil)
+		return host.executeAsyncCallGroupCallback(asyncCallGroup)
+	}
+
+	return nil
+}
+
+func (host *vmHost) executeAsyncCall(
+	asyncCall *arwen.AsyncCall,
+	syncExecutionOnly bool,
+) error {
+	execMode, err := host.determineAsyncCallExecutionMode(asyncCall)
+	if err != nil {
+		return err
+	}
+
+	if execMode == arwen.SyncExecution {
+		vmOutput, err := host.executeSyncCall(asyncCall)
+		// TODO host.executeSyncCallback() returns a vmOutput produced by executing
+		// the callback. Information from this vmOutput should be preserved in the
+		// pending AsyncCallGroup, and made available to the callback of the
+		// AsyncCallGroup (currently not implemented).
+		// Either way, host.executeSyncCallback() calls
+		// host.ExecuteOnDestContext(), which merges the vmOutput of the callback
+		// into the main output (such as OutputAccounts).
+		callbackVMOutput, callbackErr := host.executeSyncCallback(asyncCall, vmOutput, err)
+		host.finishSyncExecution(callbackVMOutput, callbackErr)
+		return nil
+	}
+
+	if syncExecutionOnly {
+		return nil
+	}
+
+	if execMode == arwen.AsyncBuiltinFunc {
+		// Built-in functions will handle cross-shard calls themselves, by
+		// generating entries in vmOutput.OutputAccounts. It is not necessary to
+		// call sendAsyncCallCrossShard(). The vmOutput produced by the built-in
+		// function, containing the cross-shard call, has ALREADY been merged into
+		// the main output by the inner call to host.ExecuteOnDestContext().
+		_, err := host.executeSyncCall(asyncCall)
+		return err
+	}
+
+	if execMode == arwen.AsyncUnknown {
+		return host.sendAsyncCallCrossShard(asyncCall)
+	}
+
+	return nil
+}
+
+func (host *vmHost) determineAsyncCallExecutionMode(asyncCall *arwen.AsyncCall) (arwen.AsyncCallExecutionMode, error) {
+	return host.determineExecutionMode(asyncCall.Destination, asyncCall.Data)
+}
+
+func (host *vmHost) determineExecutionMode(destination []byte, data []byte) (arwen.AsyncCallExecutionMode, error) {
+	runtime := host.Runtime()
+	blockchain := host.Blockchain()
+
+	// If ArgParser cannot read the Data field, then this is neither a SC call,
+	// nor a built-in function call.
+	argParser := parsers.NewCallArgsParser()
+	functionName, _, err := argParser.ParseData(string(data))
+	if err != nil {
+		return arwen.AsyncUnknown, err
+	}
+
+	code, err := blockchain.GetCode(destination)
+	if len(code) > 0 && err == nil {
```

Comment:
no need for getCode - getshardOfAddress is enough.

Reply:
Changed.